Introduce bgsave cancel

ranshid opened the pull request:
In some cases bgsave child process can run for a long time exhausting system resources. Although it is possible to kill the bgsave child process from the system shell, sometimes it is not possible allowing OS level access.

This PR adds a new subcommand to the BGSAVE command.
When user will issue `BGSAVE CANCEL`
It will do one of the 2:
1. In case a bgsave child process is currently running, the child process would be immediately killed thus terminating any save/replication full sync process.
2. In case a bgsave child process is SCHEDULED to run, the scheduled execution will be cancelled.

enjoy-binbin commented on src/rdb.c:
```diff
@@ -3645,6 +3645,18 @@ void bgsaveCommand(client *c) {
+                killRDBChild();
+                addReply(c, shared.ok);
+            } else {
+                addReplyError(c, "background save is currently not in progress");
```
```suggestion
                addReplyError(c, "Background save is currently not in progress");
```

enjoy-binbin commented on src/rdb.c:
```diff
@@ -3645,6 +3645,18 @@ void bgsaveCommand(client *c) {
+            /* BGSAVE CANCEL
+             * Terminates an in progress BGSAVE */
+            if (server.child_type == CHILD_TYPE_RDB) {
+                /* There is an ongoing save */
```
```suggestion
                /* There is an ongoing bgsave */
```

enjoy-binbin commented on tests/integration/rdb.tcl:
```diff
@@ -170,6 +170,46 @@ start_server {} {
+        
+        assert {[r bgsave cancel] eq {OK}}
+        set temp_rdb [file join [lindex [r config get dir] 1] temp-${fork_child_pid}.rdb]
+        # Temp rdb must be existed
```
```suggestion
        # Temp rdb must be deleted
```

enjoy-binbin commented on tests/integration/rdb.tcl:
```diff
@@ -170,6 +170,46 @@ start_server {} {
+        } else {
+            fail "bgsave is currently running"
+        }
+        assert_error "ERR background save is currently not in progress" {r bgsave cancel}
```
```suggestion
        assert_error "ERR Background save is currently not in progress" {r bgsave cancel}
```

enjoy-binbin reviewed (approved):
Seems like a good idea, LGTM.

hpatro commented on src/rdb.c:
```diff
@@ -3645,6 +3645,18 @@ void bgsaveCommand(client *c) {
     if (c->argc > 1) {
         if (c->argc == 2 && !strcasecmp(c->argv[1]->ptr, "schedule")) {
             schedule = 1;
+        } else if (c->argc == 2 && !strcasecmp(c->argv[1]->ptr, "cancel")) {
```
nit: should we start using `strncasecmp`?

ranshid replied:
I changed this to be a subcommand

hpatro commented on tests/integration/rdb.tcl:
```diff
@@ -170,6 +170,46 @@ start_server {} {
+        assert {[r bgsave cancel] eq {OK}}
+        set temp_rdb [file join [lindex [r config get dir] 1] temp-${fork_child_pid}.rdb]
+        # Temp rdb must be deleted
+        wait_for_condition 50 100 {
+            ![file exists $temp_rdb]
+        } else {
+            fail "bgsave temp file was not deleted after cancel"
+        }
```
Does the cleanup happen asynchronously?

ranshid replied:
it is part of the signal handler

ranshid commented on src/rdb.c:
```diff
@@ -3636,6 +3636,20 @@ void saveCommand(client *c) {
+        killRDBChild();
+        addReply(c, shared.ok);
+    } else {
+        addReplyError(c, "background save is currently not in progress");
```
```suggestion
        addReplyError(c, "Background bgsave is currently not in progress");
```

hpatro commented on src/rdb.c:
```diff
@@ -3636,6 +3636,20 @@ void saveCommand(client *c) {
 
+/* BGSAVE CANCEL */
+void bgsaveCancelCommand(client *c) {
+    /* BGSAVE CANCEL
```
Feels redundant.

ranshid replied:
fixed

ranshid commented on src/rdb.c:
```diff
@@ -3636,6 +3636,20 @@ void saveCommand(client *c) {
 
+/* BGSAVE CANCEL */
+void bgsaveCancelCommand(client *c) {
+    /* BGSAVE CANCEL
+     * Terminates an in progress BGSAVE */
```
```suggestion
    /* Terminates an in progress BGSAVE */
```

enjoy-binbin commented:
It occurred to me that maybe we also want to cancel server.rdb_bgsave_scheduled?

ranshid commented:
> It occurred to me that maybe we also want to cancel server.rdb_bgsave_scheduled?

I also thought about that during the implementation. I think the main goal was to kill the current running bgsave. Do you think this way it is currently implemented is confusing?

enjoy-binbin commented:
Looks like it could be a complement to SCHEDULE. I am ok with both.

Another question here, `kill the current running bgsave`, does it include the disk-based replication or the diskless replication intentional or unintentional?

ranshid commented:
> Looks like it could be a complement to SCHEDULE. I am ok with both.
> 
> Another question here, `kill the current running bgsave`, does it include the disk-based replication or the diskless replication intentional or unintentional?

Yes it should kill both brands

ranshid commented:
I think I can add the following logic:

if only bgsave is running kill it and ALSO cancel scheduled bgsave
if no bgsave is running BUT there is a scheduled bgsave - unschedule the scheduled BGSAVE and return O.K
if no bgsave is running OR scheduled - return an error
@enjoy-binbin WDYT?

enjoy-binbin commented:
look good to me. That's also what i thought. Let's also make sure we won't kill the replication fork? Since it is a bgsave command, i think we should let replication out or at least add a new arg for it.

ranshid commented:
> look good to me. That's also what i thought. Let's also make sure we won't kill the replication fork? Since it is a bgsave command, i think we should let replication out or at least add a new arg for it.

In AWS we encountered many cases of replication save which is taking too long and yield the system to a bad state. Sometimes the user would prefer to kill it and flush data then retry it to make the system back in good state. I do not mind adding explicit parameter to target save or replication, but I think killing the replication save is also needed

PingXie commented:
couple of comments

1. Fun time again on naming :) have we considered other options? `bgsave abort`, `bgsave stop`, `bgsave kill`, etc?  `cancel` seems to indicate some sort of gracefulness but based on the implementation it is just a process kill. So I was wondering if `abort` would be a better name?
2. I think @enjoy-binbin has a good [point](https://github.com/valkey-io/valkey/pull/757#issuecomment-2230128948). It might not be how AWS uses it but if the goal is to "stop the bleeding" quickly, I think it makes sense in general to cancel the scheduled one too.
3. Can you document exactly what this action entails as part of the PR description at the top? 

Since we are adding a new (sub)command, I consider this is a major decision so adding @valkey-io/core-team too

ranshid commented:
> couple of comments
> 
> 1. Fun time again on naming :) have we considered other options? `bgsave abort`, `bgsave stop`, `bgsave kill`, etc?  `cancel` seems to indicate some sort of gracefulness but based on the implementation it is just a process kill. So I was wondering if `abort` would be a better name?

I thought about other naming as well. IMO the kill/stop/abort is better in case we are not cancelling the scheduled save. 
In case we do cancell the scheduled save maybe cancel still has a good rational? I can also make separate sub-commands for each like:
bgsave cancel - will cancel scheduled save
bgsave kill/abort - will only stop a running bgsave
  
> 2. I think @enjoy-binbin has a good [point](https://github.com/valkey-io/valkey/pull/757#issuecomment-2230128948). It might not be how AWS uses it but if the goal is to "stop the bleeding" quickly, I think it makes sense in general to cancel the scheduled one too.

I agree. we are in full consensus here. will add it ASAP

> 3. Can you document exactly what this action entails as part of the PR description at the top?
>

Definitely.
  
> Since we are adding a new (sub)command, I consider this is a major decision so adding @valkey-io/core-team too

ranshid commented:
@enjoy-binbin @PingXie when you have the time please change the last update.
Since we cannot have both arguments and subcommands the previous change actually broke 'bgsave schedule' :) 
I had to change the format of bgsave so basically this is now a breaking change IMO

madolson commented:
@ranshid What is the value of having this be a subcommand? I feel like the previous Redis folks went overzealous about subcommands and made everything a subcommand. Given this is already an administration command, do we think should have the ability to restrict usage to cancel independently of starting a save?

zuiderkwast commented:
In which way is killing the thread non-graceful? Does it leave any junk in the file system or is it visible to the user in any way?

KILL is exposing the fact that it runs in a process, which is not necessary to iexpose in the name IMO. Besides, having both KILL and CANCEL seems excessive to me and can be confusing users. Is there use cases for both?

In which way is it a breaking change to change BGSAVE from an argument to a subcommand? Isn't it just a different way to express the same syntax?

I don't see the point of making this a completely separate command for cancelling. BGSAVE is already an admin command. BGSAVE SCEDULE looks like a subcommand or parameter. It seems logical to me to add more related operations on the form `BGSAVE [ SCEDULE | CANCEL | ... ]`. It doesn't matter to me if it's subcommands or just arguments.

PingXie commented:
>In which way is it a breaking change to change BGSAVE from an argument to a subcommand? 

We don't currently allow anything else but `schedule` to appear after `bgsave` so if someone is expecting the server to fail `bgsave cancel` they would see it succeeding instead. It is a behavior change but I also don't consider it "breaking". 

> It seems logical to me to add more related operations on the form BGSAVE [ SCEDULE | CANCEL | ... ]

+1. And also agreed that we don't need two. `cancel` works at the right abstraction level to me. 

>do we think should have the ability to restrict usage to cancel independently of starting a save?

I think `cancel` should remain an admin command.

hwware commented:
> BGSAVE [ SCEDULE | CANCEL | ... ]

I prefer this format: BGSAVE [ SCEDULE | CANCEL | KILL], and we need describe the difference between CANCEL and KILL
CANCEL represents to cancel a scheduled job , and KILL means to shutdown an on going process.

hwware commented:
@ranshid I just find one issue when I kill the child process by kill -9 child-pid command.

I run the memtier for heavy write commands, and execute bgsave on server. When server message: Background saving started by pid XXX, i manually kill the child-pid, then the memtier report the following error:

 handle error response: -MISCONF Valkey is configured to save RDB snapshots, but it's currently unable to persist to disk. Commands that may modify the data set are disabled, because this instance is configured to report errors during writes if RDB snapshotting fails (stop-writes-on-bgsave-error option). Please check the Valkey logs for details about the RDB error.

I am not sure if you have the same issue when you do the test?

ranshid commented:
> @ranshid What is the value of having this be a subcommand? 

So having this as an argument does not really provide any kind of friendly interface. There is no way I could find to provide command argument description which states that 2 arguments cannot coexist. I have no special favor for subcommands, but this was the only way I could think of in order to provide strict way to state that you can either provide no argument (BGSAVE) or BGSAVE SCHEDULE OR BGSAVE CANCEL

>In which way is killing the thread non-graceful? Does it leave any junk in the file system or is it visible to the user in any way?

We already have cases were the thread is killed and the filesystem cleanup is done as part of the signal handling. I also check this as part of the test. 

>KILL is exposing the fact that it runs in a process, which is not necessary to iexpose in the name IMO. Besides, having both KILL and CANCEL seems excessive to me and can be confusing users. Is there use cases for both?

Sure. I can change the kill and only provide the cancel option.

>In which way is it a breaking change to change BGSAVE from an argument to a subcommand?

I think that once schedule is provided as a subcommand instead of a command argument is breaking the previous docs right?
for example this is how it looks today:
```
> command docs

bgsave
summary
Asynchronously saves the database(s) to disk.
since
1.0.0
group
server
complexity
O(1)
history
3.2.2
Added the `SCHEDULE` option.
arguments
name
schedule
type
pure-token
display_text
schedule
token
SCHEDULE
since
3.2.2
flags
optional
```

and after this change it will be:
```
bgsave
summary
Asynchronously saves the database(s) to disk.
since
1.0.0
group
server
complexity
O(1)
history
3.2.2
Added the `SCHEDULE` option.
8.0.0
Refactor SCHEDULE as a subcommand
subcommands
bgsave|schedule
summary
Asynchronously saves the database(s) to disk. When an AOF rewrite is in progress and schedule the background save to r>
since
8.0.0
group
server
complexity
O(1)
bgsave|cancel
summary
Unschedule all scheduled bgsave requests.
since
8.0.0
group
server
complexity
O(1)
```

>It seems logical to me to add more related operations on the form BGSAVE [ SCEDULE | CANCEL | ... ]. It doesn't matter to me if it's subcommands or just arguments.

As I pointed before. It can be implemented as arguments and we can enforce the correct arguments were supplied (like argc <=2 ) I just think that there is no current way to correctly describe that in command docs.
If all of you feel that subcommand is not the correct way to go. I will change it back.

ranshid commented:
> @ranshid I just find one issue when I kill the child process by kill -9 child-pid command.
> 
> I run the memtier for heavy write commands, and execute "bgsave" on server. When server message: Background saving started by pid XXX, i manually kill the child-pid, then the memtier report the following error:
> 
> handle error response: -MISCONF Valkey is configured to save RDB snapshots, but it's currently unable to persist to disk. Commands that may modify the data set are disabled, because this instance is configured to report errors during writes if RDB snapshotting fails (stop-writes-on-bgsave-error option). Please check the Valkey logs for details about the RDB error.
> 
> I am not sure if you have the same issue when you do the test?
> 
> Note: bgsave kill call kill(server.child_pid, SIGUSR1);

@hwware you are killing the bgsave process with SIGKILL which is uncatchable. In this case we kill the process in more manageable (as you correctly pointed out) via SIGUSR1 which does not have the same issue.

zuiderkwast commented:
Ah, I didn't think about COMMAND DOCS. When you say docs, I think about the rendered docs on web page and man page. Anyway, COMMAND DOCS is based on the same JSON files.

> There is no way I could find to provide command argument description which states that 2 arguments cannot coexist.

In the JSON arguments, make it a "oneof" block with "option": true, no?

ranshid commented:
> Ah, I didn't think about COMMAND DOCS. When you say docs, I think about the rendered docs on web page and man page. Anyway, COMMAND DOCS is based on the same JSON files.
> 
> > There is no way I could find to provide command argument description which states that 2 arguments cannot coexist.
> 
> In the JSON arguments, make it a "oneof" block with "optional": true, no?
> 
> I'm fine with subcommand. I'm just think args should be possible too.

Let me try that. I think in arguments every top level argument must have a name and only the argument values can be enumerated, but I am far from being an expert. let me circle back on it

zuiderkwast commented:
An arg with type "pure token" has no value. There's a README for the JSON format now.

ranshid commented:
> An arg with type "pure token" has no value. There's a README for the JSON format now.

TIL

ranshid commented:
Thank you @zuiderkwast for the insights about the args :) 
I now only added the cancel argument and updated the bgsave.json. Since now I think it is no longer a breaking change I will also remove the tag.

zuiderkwast commented on src/commands/bgsave.json:
```diff
@@ -19,11 +23,24 @@
+                        "name": "schedule",
+                        "token": "SCHEDULE",
+                        "type": "pure-token",
+                        "optional": true,
```
Nice, but put optional on the oneof level instead.

The syntax is rendered from this and, although they're equivalent,

    BGSAVE [SCHEDULE | CANCEL]

looks better than

    BGSAVE <[SCEDULE] | CANCEL>

ranshid replied:
done

zuiderkwast commented on tests/integration/rdb.tcl:
```diff
@@ -170,6 +170,70 @@ start_server {} {
+
+    test {bgsave cancel response with error when no ongoing save} {
+        r config set save ""
+        # Generating RDB will take some 100 seconds
```
Copy-paste? This testcase doesn't use an ongoing save.

ranshid replied:
Yeh. I think I will merge this to the previous test

zuiderkwast reviewed (approved):
LGTM (with nits)

zuiderkwast commented on src/rdb.c:
```diff
@@ -3645,6 +3645,21 @@ void bgsaveCommand(client *c) {
+                /* There is an ongoing save */
+                serverLog(LL_NOTICE, "Background bgsave will be aborted due to user request");
+                killRDBChild();
+                addReplyStatus(c, "Background bgsave cancelled");
```
I noticed the other replies use the term "background saving" (like "Background saving scheduled").

```suggestion
                addReplyStatus(c, "Background saving cancelled");
```

zuiderkwast commented on src/rdb.c:
```diff
@@ -3645,6 +3645,21 @@ void bgsaveCommand(client *c) {
+            } else if (server.rdb_bgsave_scheduled == 1) {
+                serverLog(LL_NOTICE, "Scheduled background bgsave will be cancelled due to user request");
+                server.rdb_bgsave_scheduled = 0;
+                addReplyStatus(c, "Scheduled background bgsave cancelled");
```
```suggestion
                addReplyStatus(c, "Scheduled background saving cancelled");
```

zuiderkwast commented on src/rdb.c:
```diff
@@ -3645,6 +3645,21 @@ void bgsaveCommand(client *c) {
+                server.rdb_bgsave_scheduled = 0;
+                addReplyStatus(c, "Scheduled background bgsave cancelled");
+            } else {
+                addReplyError(c, "Background bgsave is currently not in progress or scheduled");
```
```suggestion
                addReplyError(c, "Background saving is currently not in progress or scheduled");
```

zuiderkwast reviewed (commented):
I just noticed, the formulation in replies and log messages should be "background saving" (instead of "background bgsave") to match the existing messages.

ranshid commented:
> I just noticed, the formulation in replies and log messages should be "background saving" (instead of "background bgsave") to match the existing messages.
> 
> I marked some occurrences but there are a few more.

Yeh I did use save initially. let me search and change it back

ranshid commented on src/commands/bgsave.json:
```diff
@@ -33,6 +49,12 @@
                     "const": "Background saving scheduled"
+                },
+                {
+                    "const": "Background bgsave cancelled"
```
```suggestion
                    "const": "Background saving cancelled"
```

ranshid commented on src/commands/bgsave.json:
```diff
@@ -33,6 +49,12 @@
+                    "const": "Background bgsave cancelled"
+                },
+                {
+                    "const": "Scheduled background bgsave cancelled"
```
```suggestion
                    "const": "Scheduled background saving cancelled"
```

ranshid commented on src/rdb.c:
```diff
@@ -3645,6 +3645,21 @@ void bgsaveCommand(client *c) {
+            /* Terminates an in progress BGSAVE */
+            if (server.child_type == CHILD_TYPE_RDB) {
+                /* There is an ongoing save */
+                serverLog(LL_NOTICE, "Background bgsave will be aborted due to user request");
```
```suggestion
                serverLog(LL_NOTICE, "Background saving will be aborted due to user request");
```

ranshid commented on tests/integration/rdb.tcl:
```diff
@@ -170,6 +170,64 @@ start_server {} {
+        }
+        set fork_child_pid [get_child_pid 0]
+        
+        assert {[r bgsave cancel] eq {Background bgsave cancelled}}
```
```suggestion
        assert {[r bgsave cancel] eq {Background saving cancelled}}
```

ranshid commented on tests/integration/rdb.tcl:
```diff
@@ -170,6 +170,64 @@ start_server {} {
+        } else {
+            fail "bgsave is currently running"
+        }
+        assert_error "ERR Background bgsave is currently not in progress or scheduled" {r bgsave cancel}
```
```suggestion
        assert_error "ERR Background saving is currently not in progress or scheduled" {r bgsave cancel}
```

ranshid commented on tests/integration/rdb.tcl:
```diff
@@ -170,6 +170,64 @@ start_server {} {
+        assert {[r bgsave schedule] eq {Background saving scheduled}}
+
+        # Cancel the scheduled save
+        assert {[r bgsave cancel] eq {Scheduled background bgsave cancelled}}
```
```suggestion
        assert {[r bgsave cancel] eq {Scheduled background saving cancelled}}
```

ranshid commented on tests/integration/rdb.tcl:
```diff
@@ -170,6 +170,64 @@ start_server {} {
+        assert {[r bgsave cancel] eq {Scheduled background bgsave cancelled}}
+
+        # Make sure a second call to bgsave cancel return an error
+        assert_error "ERR Background bgsave is currently not in progress or scheduled" {r bgsave cancel}
```
```suggestion
        assert_error "ERR Background saving is currently not in progress or scheduled" {r bgsave cancel}
```

ranshid commented on src/rdb.c:
```diff
@@ -3645,6 +3645,21 @@ void bgsaveCommand(client *c) {
+                killRDBChild();
+                addReplyStatus(c, "Background bgsave cancelled");
+            } else if (server.rdb_bgsave_scheduled == 1) {
+                serverLog(LL_NOTICE, "Scheduled background bgsave will be cancelled due to user request");
```
```suggestion
                serverLog(LL_NOTICE, "Scheduled background saving will be cancelled due to user request");
```

enjoy-binbin commented on tests/integration/rdb.tcl:
```diff
@@ -170,6 +170,64 @@ start_server {} {
+        wait_for_condition 50 100 {
+            [s rdb_bgsave_in_progress] == 1
+        } else {
+            fail "bgsave not done"
```
```suggestion
            fail "bgsave did not start in time"
```

enjoy-binbin reviewed (approved):
LGTM.

enjoy-binbin commented on src/rdb.c:
```diff
@@ -3645,6 +3645,21 @@ void bgsaveCommand(client *c) {
+                killRDBChild();
+                addReplyStatus(c, "Background saving cancelled");
+            } else if (server.rdb_bgsave_scheduled == 1) {
+                serverLog(LL_NOTICE, "Scheduled background saving will be cancelled due to user request");
```
i see in bgsave schedule, we did not print the server log, it only return a reply, do we need to print the cancel log in here? (maybe a reply is ok? they are not a pair so it bother me.)

enjoy-binbin replied:
lets add a server log for bgsave schedule? I feel like I'm leaning towards this.

ranshid replied:
I do not mind placing a log in case it is scheduled, but maybe it is important to distinguish the case it was explicitly requested by the user (ie BGAVE SCHEDULE) or when it was a result of BGSAVE call inside a transaction?

BTW - do you know why this support was introduced for transactions and not for scripts?

enjoy-binbin replied:
bgsave is marked with noscript flag, so it is not allowed to execute in scripts.

ranshid replied:
I know that, I wonder why it couldn't be done the same way as it is for multi/exec (means auto scheduled to be run after the script is executed)

ranshid replied:
@enjoy-binbin done.

enjoy-binbin commented on src/rdb.c:
```diff
@@ -3659,6 +3674,11 @@ void bgsaveCommand(client *c) {
     } else if (hasActiveChildProcess() || server.in_exec) {
         if (schedule || server.in_exec) {
             server.rdb_bgsave_scheduled = 1;
+            if (schedule) {
+                serverLog(LL_NOTICE, "Background saving scheduled due to user request");
+            } else {
+                serverLog(LL_NOTICE, "Background saving scheduled to run after transaction execution");
+            }
```
Maybe this is enough? i add a same line for aof, see #826
```suggestion
            serverLog(LL_NOTICE, "Background saving scheduled.");
```

ranshid replied:
I think it is clearer to explicitly state why the scheduling was done. do you think this is TMI?

enjoy-binbin replied:
yes, i think this is minor. this is a admin command, so i think it's enough to keep it simple

ranshid replied:
I agree, but for inspecting logs from distant past might be beneficial.

madolson commented:
@valkey-io/core-team Please vote on adding this new argument to cancel on ongoing save. This is useful for admins.

zuiderkwast commented:
@ranshid Do you want to prepare a doc PR, or at least create some issue for it so it's not forgotten?

soloestoy commented on src/commands/bgsave.json:
```diff
@@ -10,6 +10,10 @@
                 "Added the `SCHEDULE` option."
+            ],
+            [
+                "8.0.0",
```
we should change to 8.1.0?

zuiderkwast replied:
Ahhh yes

ranshid replied:
I will also issue a PR for that versioning fix

ranshid commented:
> @ranshid Do you want to prepare a doc PR, or at least create some issue for it so it's not forgotten?

@zuiderkwast Sure will create a doc PR

ranshid commented:
> > @ranshid Do you want to prepare a doc PR, or at least create some issue for it so it's not forgotten?
> 
> @zuiderkwast Sure will create a doc PR

@zuiderkwast  - https://github.com/valkey-io/valkey-doc/pull/182
